lightning/src/ln/msgs.rs

```diff
@@ -1956,14 +1978,27 @@ impl Readable for OnionHopData {
 		let mut payment_data: Option<FinalOnionHopData> = None;
 		let mut payment_metadata: Option<WithoutLength<Vec<u8>>> = None;
 		let mut keysend_preimage: Option<PaymentPreimage> = None;
-		read_tlv_fields!(r, {
+		let mut custom_tlvs = Vec::new();
+
+		let tlv_len = BigSize::read(r)?;
+		let rd = FixedLengthReader::new(r, tlv_len.0);
+		decode_tlv_stream_with_custom_tlv_decode!(rd, {
 			(2, amt, required),
 			(4, cltv_value, required),
 			(6, short_id, option),
 			(8, payment_data, option),
 			(16, payment_metadata, option),
 			// See https://github.com/lightning/blips/blob/master/blip-0003.md
 			(5482373484, keysend_preimage, option)
+		}, |msg_type: u64, msg_reader: &mut FixedLengthReader<_>| -> Result<bool, DecodeError> {
+			if msg_type < 1 << 16 { return Ok(false) }
```

Comment:
Currently in this PR, as long as the custom TLVs are within the custom type number range, we'll succeed deserializing and provide the bytes to the user, regardless of whether the user actually recognizes these type numbers. I did this as a simple way to start, but IIUC we should actually fail deserialization if it's unknown + even. So should I add some way for the user to provide a custom handler here? Or at the least let them provide recognized type numbers?

Reply:
Yea sorry was going to come back to this.

> accept everything at the deserialization stage and then require the user call some claim_payment_with_known_custom_tlvs thing if they're even?

As in the normal `claim_funds` would fail if the attempted payment had even custom TLVs on it? And then `claim_payment_with_known_custom_tlvs` would just be the same thing except that it wouldn't fail, and would only be used when the user successfully recognized all the even TLVs?

Also I hadn't gotten around to it yet but @wpaulino had mentioned offline adding another `channelmanager::FailureCode` variant for the `invalid_onion_payload` error so the user could `ChannelManager::fail_htlc_backwards_with_reason` if they failed to recognize an even TLV - was planning to do this as well

Reply:
Added functionality for everything mentioned here